torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -52,6 +53,79 @@
 zip = strict_zip
 
 
+class CompilerWrapper:
+    """
+    A wrapper around the inputs and outputs to the compiler_fn. We separate these into two parts:
+
+    1. The prologue, which edits the input to the compiler_fn(flat_fn, flat_args, etc)
+    2. The epilogue, which edits the outputs of the compiler_fn (compiled_fn, real arguments)
+
+    Each wrapper below should be implemented as a CompilerWrapper, so that we can facilitate
+    caching on the compiled output, and re-wrapping the output via epilogues.
+    Extra metadata that is needed to compute pre or post compile can be passed in via attributes.
+    """
+
+    def pre_compile(
+        self,
+        flat_fn,
+        flat_args: List[Tensor],
+        aot_config: AOTConfig,
+        *,
+        fw_metadata: ViewAndMutationMeta,
```

Comment:
Just another callout - right now, we have:

(1) multiple transformations (each of which will eventually get its own `CompilerWrapper` according to this PR)
(2) each transformation requires some metadata

And some of that metadata is shared across transformations (like `aot_config` and `fw_metadata`, which you're uniformly passing into every instance of `CompilerWrapper`, while other metadata is specific to a certain transformation (like `indices_of_inps_to_detach`).

I definitely wouldn't block your current refactor on this extra refactor, but I do wonder if we should aim to go more in either of those directions: either each `CompilerWrapper` gets exactly the metadata it cares about and nothing is shared, or we glob all metadata into a single shared object that is plumbed around everywhere (we definitely lean more in this direction today)

Reply:
Hmm this is a bit tricky to get right I think: I don't really think it's possible to make each CompilerWrapper completely independent, simply because the specific fields that are shared are also modified by each CompilerWrapper's pre_compile functions, so they're not actually independent. 

Having a single metadata structure represent *all* of the shared info would definitely be nice, and also is pretty much what a cache entry would look like, so I might end up doing that refactor when I actually create cache entries(basically, passing along a single data structure with all the fields shared between CompilerWrappers). 

Though it's also kind of nice, code readability wise, for CompilerWrappers to have statically defined parameters for fields like `indices_of_inps_to_detach` instead of a single "metadata" field: it means if you change metadata it's easier to find out which wrappers consume the data, instead of having to hunt through each usage in the code. 

I think the compromise here might be that there is a single object we plumb through aot autograd, but the compilerWrappers themselves still just take individual arguments: note that none of these CompilerWrapper objects should need to be directly plumbed outside the function they're defined in: in a perfect world or end state, we should just have a series of pre compile steps, followed by actual compilation (aot_dispatch_base/autograd), followed by a series of post compile steps.